### PR TITLE
Add functionality to get vision data for the Visual Mesh

### PR DESCRIPTION
**Pull request by ysims**
This PR 

- adds the ball directly into our repo from Webots source, and modify it to add segmentation (red)
- modifies the field proto, so field lines and green grass are wrapped in separate `Solid`s with segmentation (white and green respectively)
- modifies the goal proto, so all posts are wrapped in a `Solid` with segmentation (yellow) - small poles and large poles are separated into different shades of yellow.
- modifies NUgus proto so it has segmentation (blue)
- adds the Darwin proto and its controller and modifies the Darwin proto to have segmentation (blue)
- adds more robots: Darwin-OP3 and Bit-Bot's Wolfgang robot
- Adds a world specifically for collecting vision data
- New `collect_vision_data` controller which collects vision data.

### CODE REVIEW

File: controllers/collect_vision_data/collect_vision_data.cpp [hunk 1]
```diff
@@ -91,6 +91,15 @@ int main(int argc, char** argv) {
     right_camera->enableRecognitionSegmentation();
     left_camera->enableRecognitionSegmentation();
 
+    int camera_width = left_camera->getWidth();
+    int camera_height = left_camera->getHeight();
+    double horizontal_fov = left_camera->getFov();
+    double vertical_fov = 2 * std::atan(std::tan(horizontal_fov * 0.5) * ((double) camera_height / camera_width));
```
Comment (TrentHouliston) [inline controllers/collect_vision_data/collect_vision_data.cpp:97]: Need some clang tidy in here to protect you from these C style casts!
```suggestion
    double vertical_fov = 2 * std::atan(std::tan(horizontal_fov * 0.5) * (double(camera_height) / camera_width));
```

File: controllers/collect_vision_data/collect_vision_data.cpp [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+#include <iostream>
+#include <random>
+#include <sstream>
+#include <webots/Camera.hpp>
+#include <webots/Motor.hpp>
+#include <webots/PositionSensor.hpp>
+#include <webots/Robot.hpp>
+#include <webots/Supervisor.hpp>
+
+#include "yaml-cpp/yaml.h"
```
Comment (Bidski) [inline controllers/collect_vision_data/collect_vision_data.cpp:35]:
```suggestion
#include <yaml-cpp/yaml.h>
```

File: controllers/collect_vision_data/collect_vision_data.cpp [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+std::string padLeft(int number, int width) {
+    std::stringstream ss;
+    ss << std::setw(width) << std::setfill('0') << number;
+    return ss.str();
+}
+
+const int QUALITY = 100;
```
Comment (Bidski) [inline controllers/collect_vision_data/collect_vision_data.cpp:43]: Should probably comment this

File: controllers/collect_vision_data/collect_vision_data.cpp [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    // X ranges from -5.4 - 5.4, the positive value on the left hand side when looking
+    // from red
+    // Y ranges from -3.8 - 3.8, the positive value on the red goal side
+    // Z should be set to 0.51 to be on the ground (and add a bit of noise)
+    std::uniform_real_distribution<> xDistrib(-5.4, 5.4);
+    std::uniform_real_distribution<> yDistrib(-3.8, 3.8);
+    std::uniform_real_distribution<> zDistrib(0.46, 0.56);
+
+    // Move servos in the range [-PI/2, PI/2]
+    std::uniform_real_distribution<> servoDistrib(-M_PI * 0.5, M_PI * 0.5);
```
Comment (Bidski) [inline controllers/collect_vision_data/collect_vision_data.cpp:122]:
```suggestion
    std::uniform_real_distribution<> servoDistrib(-M_PI_2, M_PI_2);
```

**Review by Bidski — CHANGES_REQUESTED** [review]
Need to delete the binary files from the repo as well (`nugus_controller_test`, `controllers/vision_test/vision_test`, `controllers/darwin_controller/darwin_controller`)

File: controllers/collect_vision_data/collect_vision_data.cpp (hunk 3 above)
Comment (TrentHouliston) [reply controllers/collect_vision_data/collect_vision_data.cpp:43]: Also, is this a 100 quality jpeg? That's crazy talk!
My guess is for the actual competition you'll be down at like 75 or less

Comment (Bidski) [reply controllers/collect_vision_data/collect_vision_data.cpp:43]: Its also used for PNGs I think

File: protos/robot/NUgus.proto [hunk 5]
```diff
@@ -0,0 +1,1256 @@
+#VRML_SIM R2021a utf8
+# license: Apache License 2.0
+# license url: http://www.apache.org/licenses/LICENSE-2.0
+# This is a proto file for Webots for the NUgus
+# Extracted from: NUgus.urdf
+
+PROTO NUgus [
+  field  SFVec3f     translation     0 0 0
+  field  SFRotation  rotation        1 0 0 0                         # Make up be the positive Z axis
```
Comment (Bidski) [inline protos/robot/NUgus.proto:9]: Comment makes no sense now
```suggestion
  field  SFRotation  rotation        1 0 0 0 
```

**Review by Bidski — CHANGES_REQUESTED** [review]

File: controllers/collect_vision_data/collect_vision_data.cpp (hunk 3 above)
Comment (TrentHouliston) [reply controllers/collect_vision_data/collect_vision_data.cpp:43]: PNGs are just always 100% quality, they're a lossless compression

Comment (Bidski) [reply controllers/collect_vision_data/collect_vision_data.cpp:43]: Just had a look at the docs. They ignore the quality parameter for PNGs (as they should).

**Review by JosephusPaye — COMMENTED** [review]
Just tested and this now LGTM. I've attached the images and lens params I used and the corresponding visual mesh output for review. 

[images_lens_drawn_mesh.zip](https://github.com/NUbots/NUWebots/files/6506342/images_lens_drawn_mesh.zip)

 @TrentHouliston do those meshes look right? We wanna merge this and start getting data for training.

**Comment by Bidski** [comment]
@LachlanCourt did you mean to remove all of that code and the config file?

**Comment by ysims** [comment]
> @LachlanCourt did you mean to remove all of that code and the config file?

We're moving to use the new teleport controller, but the world file (worlds/vision_collection.wbt) needs to be changed still.

**Comment by LachlanCourt** [comment]
> @LachlanCourt did you mean to remove all of that code and the config file?

The teleport_master_controller handles setting the position and rotation now so this controller doesn't need to set it anymore, just needs to grab the current position and rotation when it is set, and then grab the image data

**Comment by ysims** [comment]
> > @LachlanCourt did you mean to remove all of that code and the config file?
> 
> The teleport_master_controller handles setting the position and rotation now so this controller doesn't need to set it anymore, just needs to grab the current position and rotation when it is set, and then grab the image data. @ysims do you want me to set the world file up so that it will run with the master controller straight away?

Yes, that will need to be done and tested so that it still works as intended. Make sure you are changing `vision_collection.wbt` not `kid.wbt`.

**Comment by LachlanCourt** [comment]
Okay I think we done :sweat_smile:

**Review by Bidski — APPROVED** [review]
LGTM

**Review by LachlanCourt — APPROVED** [review]
Whoops I dismissed that review by merging main 🤦

**Pull request merged**